Vectra AI with accounts

Vectra-ODuda opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Adds Vectra accounts related functions + improves incoming mapper

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [x] 6.5.0

## Does it break backward compatibility?
   - [x] Yes
       - Further details: Some mapper attributes have been removed (as duplicates) or renamed
   - [ ] No

## Must have
- [x] Tests
- [x] Documentation

idovandijk commented:
Thank you @Vectra-ODuda for the contribution. While you're waiting for us to start the review, can you please elaborate on the addition of the "accounts"? What is the improvement you're adding in this PR?
The information will help with the review process.

YuvHayun commented on Packs/Vectra_AI/Integrations/VectraDetect/VectraDetect.yml:
```diff
@@ -65,21 +65,31 @@ configuration:
-  defaultvalue: host.threat:>=50 AND host.certainty:>=50
+  defaultvalue: host.threat:>=50
+
+- display: Accounts fetch query
```
please add a note about this in the RN.

Vectra-ODuda replied:
Hi @YuvHayun , regarding the CommonTypes RN, I've asked the question to @kgal-pan but we were unsure how what we need to do.
I'll make the changes.
With that change, how can I put a strong dependency on the minimum CommonTypes pack version to be mandatory for my Pack version ?
Thx in advance

Vectra-ODuda replied:
Done regarding the Vectra AI Pack RN

YuvHayun replied:
Actually what I meant is, that we should add the new param (accounts fetch query) to the RN.

Vectra-ODuda replied:
Oh ok, makes sense, let me do it also.

YuvHayun replied:
So just making sure, everything is alright here? no more questions regarding this comment?

Vectra-ODuda replied:
Still the strong dependency question now that I've updated the CommonTypes version, is there a way to do that on my pack_metadata file ?

YuvHayun replied:
Not sure I understand your questions, can you please elaborate?

Vectra-ODuda replied:
Sure, let me rephrase. As I was forced to update the "Common Types" pack to allow some of my incident type to be used with Common Types incident fields, I've added a dependency on this pack on my pack_metadata file.
However, just adding the pack name might not be sufficient if use an older version not having my changes.
So I'm asking if there is a way to say Vectra AI Pack v1.1.0 has a dependency on Common Types (minimum version 3.2.47) that way I'm sure my pack will works as expected or the mapping might failed.

YuvHayun replied:
There's no need to mention that anywhere - the dependencies list is calculated automatically and will include the min version for common types.
When users will upgrade your pack it'll also upgrade Common types as well.

Vectra-ODuda replied:
Ok, I have removed the dependency as suggested

YuvHayun commented on Packs/Vectra_AI/Integrations/VectraDetect/VectraDetect.py:
```diff
@@ -737,9 +737,17 @@ def extract_detection_data(detection: Dict[str, Any]) -> Dict[str, Any]:
 
+    summary = detection.get('summary')
+    if summary:
+        description = summary.get('description')
+        dst_ips = summary.get('dst_ips')
+        dst_ports = summary.get('dst_ports')
```
```suggestion
        description = summary.get('description', "")
        dst_ips = summary.get('dst_ips', "")
        dst_ports = summary.get('dst_ports', "")
```

Vectra-ODuda replied:
Hi @YuvHayun, so you prefer to have an empty string than a None value. I'll change that

Vectra-ODuda replied:
Done

YuvHayun reviewed (changes requested):
Hey @Vectra-ODuda, thx for your contribution
Small comments from my side, you can see them below.
Also, update the CommonType RN as well please.

idovandijk commented on Packs/CommonTypes/IncidentFields/incidentfield-ExtraHop_Detection_ID.json:
```diff
@@ -24,7 +24,8 @@
  "group": 0,
  "hidden": false,
  "associatedTypes": [
-  "ExtraHop Detection"
```
This one is related to the ExtraHop integration, I agree it shouldn't be in the CommonTypes pack.
Please see if there's a different field you can use for Vectra Detection ID or you may create a new field in your pack.

Vectra-ODuda replied:
Hi @idovandijk , indeed I saw that this field related to ExtraHop but being part of the CommonTypes, I was initialy thinking of updating the description to be more generic but also the .json file name to be usable by everyone, not only ExtraHop as quite confusing.
I'll see want I can use instead.

idovandijk replied:
Yes, let's not touch the existing field as its name already suggests it's for another integration :)

Vectra-ODuda replied:
Hi @idovandijk , so far, the “ExtraHop” usage is only visible under the `description`, neither on the `name` or `id` or `cliName` directly. I’ve put aside the `associatedTypes` as this one can have multiples values for any fields.
This is only visible on the .json file name.
Are you OK if I simply update the description to be more generic ? So it could be used by any other integration.
Also should I change the file name (with a git move) or should I left it intact ?
Bonus question: Should I do the same changes on the two others "ExtraHop" incident fields `Detection Update Time`and `Detection End Time` ?
Thx in advance

idovandijk reviewed (changes requested):
Nice job! I only found 1 thing that we need to change :)

YuvHayun commented on Packs/Vectra_AI/Integrations/VectraDetect/VectraDetect.py:
```diff
@@ -737,9 +737,17 @@ def extract_detection_data(detection: Dict[str, Any]) -> Dict[str, Any]:
     source_host = detection.get('src_host')
     source_host_id = source_host.get('id') if source_host else None
 
+    summary = detection.get('summary')
+    if summary:
+        description = summary.get('description', "")
+        dst_ips = summary.get('dst_ips', "")
+        dst_ports = summary.get('dst_ports', "")
+
```
```suggestion
    summary = detection.get('summary')
    data_dict = {}
    if summary:
        if detection.get('category'):
            data_dict['Category'] = detection.get('category')
        if summary.get('description', ""):
            data_dict['description'] = summary.get('description', "")
        if summary.get('dst_ips', ""):
            data_dict['dst_ips'] = summary.get('dst_ips', "")
        if summary.get('dst_ports', ""):
            data_dict['dst_ports'] = summary.get('dst_ports', "")
```
Taking back what I said, sorry!

Vectra-ODuda replied:
Hi @YuvHayun , if you're OK I've done that, it's quicker and allow me to see all value into a single dict (with ordered keys) and won't change the others parts. Because `summary` key is always present, so it's mainly for PyPy to be happy.
```
summary = detection.get('summary')
if summary:
    description = summary.get('description')
    dst_ips = summary.get('dst_ips')
    dst_ports = summary.get('dst_ports')
else:
    description = dst_ips = dst_ports = None
```

Vectra-ODuda replied:
Thx for your `remove_empty_elements` trick on your committed fix @YuvHayun.
I will use that one more often :)

YuvHayun commented on Packs/Vectra_AI/Integrations/VectraDetect/VectraDetect.py:
```diff
@@ -737,9 +737,17 @@ def extract_detection_data(detection: Dict[str, Any]) -> Dict[str, Any]:
+        dst_ips = summary.get('dst_ips', "")
+        dst_ports = summary.get('dst_ports', "")
+
     return common_extract_data(detection) | {
         'Category'            : detection.get('category'),                                                # noqa: E203
```
```suggestion
    return common_extract_data(detection) | data_dict  # noqa: E203
```

Vectra-ODuda replied:
See suggested committed fix above for consistency/readability if you're ok

YuvHayun commented on Packs/Vectra_AI/Integrations/VectraDetect/VectraDetect.py:
```diff
@@ -737,9 +737,17 @@ def extract_detection_data(detection: Dict[str, Any]) -> Dict[str, Any]:
     return common_extract_data(detection) | {
         'Category'            : detection.get('category'),                                                # noqa: E203
-        'Description'         : detection.get('description'),                                             # noqa: E203
+        'Description'         : description,                                                              # noqa: E203
+        'DestinationIPs'      : dst_ips,                                                                  # noqa: E203
+        'DestinationPorts'    : dst_ports,                                                                # noqa: E203
```
```suggestion
```

Vectra-ODuda replied:
See suggested committed fix above for consistency/readability if you're ok

YuvHayun reviewed (changes requested):
added fixes, they should also fix your Lint & unit tests problems.

YuvHayun commented on Packs/Vectra_AI/pack_metadata.json:
```diff
@@ -3,16 +3,26 @@
     "created": "2022-06-27T10:00:00Z",
-    "description": "This content pack allows to create incidents based on Vectra Accounts/Detections/Hosts objects.",
+    "currentVersion": "1.1.0",
+    "dependencies": {
+        "CommonTypes": {
+            "display_name": "Common Types",
+            "mandatory": true
+        }
+    },
```
```suggestion
```
Due to my dependencies answer on the other comment, you can delete that.

Vectra-ODuda replied:
Done as suggested

Vectra-ODuda commented:
Hi, how can I fix the mapper backward compatibly error message on CircleCI as my pack is on Beta ?
Is there is missing flag on the pack_metadata file ?

> ERROR]: Packs/Vectra_AI/Classifiers/classifier-Vectra_Detect_-_Incoming_Mapper.json: [MP107] - Some incident fields were removed from the mapper, The removed fields: {'Vectra Host': {'Vectra Threat Score', 'Is Active', 'Vectra Certainty Score'}, 'Vectra Detection': {'State', 'occurred'}, 'dbot_classification_incident_type_all': {'Source IP', 'External Sub Category Name', 'Detection URL', 'Alert ID', 'External Start Time', 'External Category Name', 'Incident Link', 'Description', 'External End Time'}, 'Vectra Account': {'Is Active', 'Vectra Threat Score', 'Vectra Certainty Score'}}

Some of them as moved to "type_all" instructions, some have been renamed.
Thx in advance

YuvHayun commented:
> Hi, how can I fix the mapper backward compatibly error message on CircleCI as my pack is on Beta ? Is there is missing flag on the pack_metadata file ?
> 
> > ERROR]: Packs/Vectra_AI/Classifiers/classifier-Vectra_Detect_-_Incoming_Mapper.json: [MP107] - Some incident fields were removed from the mapper, The removed fields: {'Vectra Host': {'Vectra Threat Score', 'Is Active', 'Vectra Certainty Score'}, 'Vectra Detection': {'State', 'occurred'}, 'dbot_classification_incident_type_all': {'Source IP', 'External Sub Category Name', 'Detection URL', 'Alert ID', 'External Start Time', 'External Category Name', 'Incident Link', 'Description', 'External End Time'}, 'Vectra Account': {'Is Active', 'Vectra Threat Score', 'Vectra Certainty Score'}}
> 
> Some of them as moved to "type_all" instructions, some have been renamed. Thx in advance

If @idovandijk approve this changes, we'll force merge this.

idovandijk commented:
> > Hi, how can I fix the mapper backward compatibly error message on CircleCI as my pack is on Beta ? Is there is missing flag on the pack_metadata file ?
> > > ERROR]: Packs/Vectra_AI/Classifiers/classifier-Vectra_Detect_-_Incoming_Mapper.json: [MP107] - Some incident fields were removed from the mapper, The removed fields: {'Vectra Host': {'Vectra Threat Score', 'Is Active', 'Vectra Certainty Score'}, 'Vectra Detection': {'State', 'occurred'}, 'dbot_classification_incident_type_all': {'Source IP', 'External Sub Category Name', 'Detection URL', 'Alert ID', 'External Start Time', 'External Category Name', 'Incident Link', 'Description', 'External End Time'}, 'Vectra Account': {'Is Active', 'Vectra Threat Score', 'Vectra Certainty Score'}}
> > 
> > 
> > Some of them as moved to "type_all" instructions, some have been renamed. Thx in advance
> 
> If @idovandijk approve this changes, we'll force merge this.

@Vectra-ODuda can you please declare these changes as breaking backward compatibility in the release notes? After that I will approve the PR

Vectra-ODuda commented:
> > > Hi, how can I fix the mapper backward compatibly error message on CircleCI as my pack is on Beta ? Is there is missing flag on the pack_metadata file ?
> > > > ERROR]: Packs/Vectra_AI/Classifiers/classifier-Vectra_Detect_-_Incoming_Mapper.json: [MP107] - Some incident fields were removed from the mapper, The removed fields: {'Vectra Host': {'Vectra Threat Score', 'Is Active', 'Vectra Certainty Score'}, 'Vectra Detection': {'State', 'occurred'}, 'dbot_classification_incident_type_all': {'Source IP', 'External Sub Category Name', 'Detection URL', 'Alert ID', 'External Start Time', 'External Category Name', 'Incident Link', 'Description', 'External End Time'}, 'Vectra Account': {'Is Active', 'Vectra Threat Score', 'Vectra Certainty Score'}}
> > > 
> > > 
> > > Some of them as moved to "type_all" instructions, some have been renamed. Thx in advance
> > 
> > 
> > If @idovandijk approve this changes, we'll force merge this.
> 
> @Vectra-ODuda can you please declare these changes as breaking backward compatibility in the release notes? After that I will approve the PR

This is now done. Let me know if it's ok for you.

idovandijk reviewed (approved):
Great job!

Vectra-ODuda commented:
Hi @idovandijk, I've pushed the modifications regarding the following ExtraHop Incident fields `Detection ID`, `Detection End Time` and `Detection Update Time` to be clean by removing description (as on all generic fields) also with a RN addition regarding the descriptions deletion.
Let me know if it's OK for you

Vectra-ODuda commented:
> Hi @idovandijk, I've pushed the modifications regarding the following ExtraHop Incident fields `Detection ID`, `Detection End Time` and `Detection Update Time` to be clean by removing description (as on all generic fields) also with a RN addition regarding the descriptions deletion. Let me know if it's OK for you
> 
> [Edit] The ciclici: Run validations is not happy on the "Common Types" pack with this error on the 3 changes `[IF115] - The unsearchable key in indicator and incident fields should be set to true.`however this attribute is set to false on any incident fields of this Pack. Maybe you (Cortex Team) should handle that :)

Hi @idovandijk , @YuvHayun , any news on this one ? this is the only blocker I see for this PR.
Thx in advance

YuvHayun commented on Packs/Vectra_AI/ReleaseNotes/1_1_0.md:
```diff
@@ -0,0 +1,42 @@
+**/!\ Caution: This version breaks backward compatibility due to removed/renamed incident fields when Feching Incidents.**  
+**Please find more details on the Mapper changes. /!\\**
```
```suggestion
**/!\ Caution: This version breaks backward compatibility due to removed/renamed incident fields when Feching Incidents for ***classifier-Vectra_Detect_-_Incoming_Mapper*** mapper.**  
**Please find more details on the Mapper changes. /!\\**
```

Vectra-ODuda replied:
done

YuvHayun commented:
Hi @Vectra-ODuda, I added 1 comment to fix the RN warning.
Also, waiting for @idovandijk input about the pr.

Vectra-ODuda commented:
> Hi @Vectra-ODuda, I added 1 comment to fix the RN warning.
> Also, regarding the unsearchable keys - it's up to you to choose if you want these fields to be searchable or not.

Hi @YuvHayun , regarding the unsearchable key, the warning occurs on the "Common Types" pack validation, not on the Vectra AI one. As this pack is mainly maintained by Palo Alto, I'm not sure why the validator complains about it as I've just removed the `description` attribute and that I've left the `unsearchable` key untouched.

idovandijk commented:
> > Hi @Vectra-ODuda, I added 1 comment to fix the RN warning.
> > Also, regarding the unsearchable keys - it's up to you to choose if you want these fields to be searchable or not.
> 
> Hi @YuvHayun , regarding the unsearchable key, the warning occurs on the "Common Types" pack validation, not on the Vectra AI one. As this pack is mainly maintained by Palo Alto, I'm not sure why the validator complains about it as I've just removed the `description` attribute and that I've left the `unsearchable` key untouched.

Hi @Vectra-ODuda, since the fields failing the build are common ones and you only changed the incident associations, we will force merge this build (there's no issue that you need to fix here)

YuvHayun commented:
@Vectra-ODuda Merging this pr, we'll take care of the force merge.
Thx for your contribution!
